### ISSUE TITLE
Doc suggstion (NixOS)

### ISSUE THREAD

**Issue by grstearns** [democratic-csi/democratic-csi#421]
I was having issues deploying an iscsi pvc via k3s in NixOS and tracked it down to the fact that the open-iscsi config is [kind of wonky](https://discourse.nixos.org/t/configure-iscsi/50773). 

The first error was that it could not find `iscsiadm`:
> rpc error: code = Internal desc = {"code":127,"stdout":"","stderr":"/usr/bin/env: 'iscsiadm': No such file or directory\n","timeout":false}

I had it in my configuration.nix and was able to verify that it was installed

configuration.nix:
```nix
environment.systemPackages = with pkgs; [
  nfs-utils
  openiscsi
  lsscsi
  sg3_utils
  multipath-tools
];
```
```sh
$ iscsiadm --version
iscsiadm version 2.1.10
```

The workaround was to link it in a place that it could find it:
```sh
sudo ln -s $(which iscsiadm) /usr/bin/iscsiadm
```

Once that was in place, I got a second error:

> MountVolume.MountDevice failed for volume "pvc-8064ba03-d980-497c-8a64-88ee9145d48b" : rpc error: code = Internal desc = {"code":20,"stdout":"Logging in to [iface: default, target: iqn.2005-10.org.freenas.ctl:csi-pvc-8064ba03-d980-497c-8a64-88ee9145d48b-clustera, portal: 10.111.1.50,3260]\n","stderr":"iscsiadm: iscsid is not running. Could not start it up automatically using the startup command in the iscsid.conf iscsid.startup setting. Please check that the file exists or that your init scripts have started iscsid.\niscsiadm: can not connect to iSCSI daemon (111)!\niscsiadm: Could not login to [iface: default, target: iqn.2005-10.org.freenas.ctl:csi-pvc-8064ba03-d980-497c-8a64-88ee9145d48b-clustera, portal: 10.111.1.50,3260].\niscsiadm: initiator reported error (20 - could not connect to iscsid)\niscsiadm: Could not log into all portals\n","timeout":false}

Ultimately, the fix was to not just install, but enable open-iscsi.

configuration.nix:
```nix
   services.openiscsi = {
    enable = true;  # Enable openiscsi daemon
    name = "iqn.2024-09.tld.org.your:nixos-hostname;" # Any valid iqn will work, we just need this for the service to start

    discoverPortal = "10.111.1.50";   # TrueNAS IP - might not be needed
  };
```
Again, this was tested with k3s (unstable - currently v1.31.0+k3s1), but might also see issues with rke / rke2. I will likely write up a PR in a bit, but I first wanted to get those error messages someplace where someone in need might find them (this was basically my entire afternoon).

**Comment by travisghansen** [comment]
I have some familiarity with nixos and that all sounds sane. You can set an env var to explicitly set the path to the iscsiadm binary (https://github.com/democratic-csi/democratic-csi/blob/master/docker/iscsiadm is the wrapper used).

What do you need in a PR? Just documentation or something else?

**Comment by travisghansen** [comment]
FYI discovery portal should not be required.

**Comment by grstearns** [comment]
Just docs. I know I can't be the only person who will ever run into this

**Comment by travisghansen** [comment]
Ok cool thanks! The only other thing that could be a little odd is the locations of the ‘db’ directories. /etc/iscsi and /var/lib/iscsi…are those in strange locations in nixos?

**Comment by grstearns** [comment]
I'm still getting my feet wet with nix, but it seems that they do seem to keep /etc and /var intact. At most, they'll link something in there to the store. I double checked and those dirs are solid.

The iscsiadm thing kind of threw me since it really did end up someplace weird:

```sh
which iscsiadm
/run/current-system/sw/bin/iscsiadm
```

**Comment by travisghansen** [comment]
Well, even that location is more of a virtual thing…the actual location is in the store somewhere.  Where did you put the ln -s command? Is it run on every boot or something?

My experience with nixos is even stranger as I use it to create an iot-like distro for devices in the wild…the same image is used across a bunch of devices so a bunch of the configuration has to be very ephemeral but other little bits need to be unique per-device etc. It’s a wild beast!

**Comment by grstearns** [comment]
Right, that too gets flung into the store. But it should be stable referencing "current-system" (at least I hope).

That command was just a one-off. I did a couple rebuilds and it seems to be ok. But setting the path in a variable feels way less hacky.

Yeah, this really has been a ride. I was also playing around with Suse microos and it seems cool, but there's just so much more info about nixos.

**Comment by travisghansen** [comment]
I think that will work yes, I would probably codify it in configuration but one-off should work.

**Comment by travisghansen** [comment]
I think the best way to handle this is to add the nixos path(s) to all the relevant wrapper scripts in this codebase. I can add that in the next release.

**Comment by starcraft66** [comment]
FYI, I am using `democratic-csi` on NixOS as well and saw this issue so I'll chime in quickly.

@grstearns The reason why you can't just add `iscsiadm` to your environment and attempt to run it is because it is merely a client tool that relies on the `iscsid` service, which gets enabled when you enable the `openiscsi` modules.

Portal discovery is indeed not needed.

Here's what I enable on my control-plane nodes to get nfs and iscsi support working: https://github.com/starcraft66/infrastructure/blob/d2b21b2fb7aefce91dfbcdf6c47bddf40d2ba416/nixos/modules/kubernetes/worker/kubelet.nix#L11-L18

I was in a hurry to get things minimally working back when I set this up so I was incredibly lazy and [used](https://github.com/starcraft66/infrastructure/blob/d2b21b2fb7aefce91dfbcdf6c47bddf40d2ba416/k8s-235/apps/democratic-csi/base/resources/values-iscsi.yaml#L98-L102) `/run/current-system/sw/bin` as a crutch to get iscsiadm mounted into the container as well but there is a better (perhaps still a bit abusive) solution for this which I'll take the time to point out: You can add something like
```nix
environment.etc."iscsiadm".source = pkgs.iscsiadm;
```
which would end up dumping iscsiadm in `/etc/static/iscsiadm/bin/iscsiadm`, a predictable location that won't change regardless of your current system as long as that line is part of your config ([source](https://search.nixos.org/options?channel=24.05&show=environment.etc.%3Cname%3E.source&from=0&size=50&sort=relevance&type=packages&query=etc.)).

**Comment by travisghansen** [comment]
I think I'll just add `/run/current-system/sw/bin` to the `PATH` of all those relevant scripts. Should be included in the next release.
